src/Chapter3.hs

```diff
@@ -374,6 +382,40 @@ after the fight. The battle has the following possible outcomes:
 
 -}
 
+data Fighter' = 
+    KnightF  { fighterHealth :: Int
+             , fighterAttack :: Int
+             , fighterGold :: Int
+             }
+
+  | MonsterF { fighterHealth :: Int
+             , fighterAttack :: Int
+             , fighterGold :: Int
+             }
+  deriving (Show)
+
+-- The instructions aren't totally clear, so I'm making some assumptions:
+--  1. A fight is a single round in which knight attacks monster and monster attacks knight
+--  2. The fight is to the death, so loser has health < 0 at the end of the round
+--  3. If both knight and monster have health >=0 neither wins
+
+-- fight consists of two attacks:
+--  1. knight attacks monster, resulting in m1 (monster with updated health)
+--  2. m1 attacks knight (resulting in knight with updated health)
+fight' :: Fighter' -> Fighter' -> Int
+fight' k m = result (attack m1 k) m1
+  where
+    m1 = attack k m
+    result k' m'
+      | fighterHealth k' > 0 && fighterHealth m' < 0  = fighterGold k' + fighterGold m'
+      | fighterHealth k' < 0 && fighterHealth m' > 0  = -1
+      | otherwise                                   = fighterGold k'
+
+-- Attack is half a round of the fight: fighter 1 attacks fighter 2, and the return value
+-- is fighter 2 with its health reduced by fighter 1's attack.
+attack :: Fighter' -> Fighter' -> Fighter'   
+attack f1 f2 = f2 { fighterHealth = fighterHealth f2 - (fighterAttack f1) }
+
```

Comment:
That is a nice solution!

I also see that the attack function can be inlined into the `fight'` function itself.
That means that you could just check is `fighterHealth moster < fighterAttack knight` and vice versa 🙂 
But this is just a suggestion :)